etag can be passed in the if-none-match in the header on getobject api

vishalsanfran opened the pull request:
Closes #5967

## Change Description
Add an optional [If-None-Match: "ETag"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match#syntax) header to getObject requests. When an ETag is passed, if that ETag is current for the object on that path then return 304 Not Modified.

If the ETag is not current, return the object as it was being originally done. This ensures that this is not a breaking change.

### Background

I had earlier created a PR for the same issue https://github.com/treeverse/lakeFS/pull/6440. But, there was some back and forth and it got closed. Hoping to get it done this time.

### New Feature

In the clients, the `getObject` API will now support an optional `ifNoneMatch` field.
For eg in Java:
getObject(repository, ref, path).range(range).**ifNoneMatch(ifNoneMatch)**.presign(presign).execute();

### Testing Details

Added a unit test that validates that a 304 is returned when the eTag is the same as that in the request.

### Breaking Change?

If-None-Match is an optional parameter and existing clients will need to be updated to pass this parameter. So, it should not be a breaking change.

##Contact
I'll do my best, but Please contact me on vishalkk@gmail.com in case I'm not responding to comments.

arielshaqed reviewed (changes requested):
Thanks for this PR!  Please see comments inside: I am optimistic about backwards compatibility with the new HTTP status code return, no action required there right now.

The required change is:
- Test getObject with If-None-Match that doesn't match.

Requested changes, which will not block, are:
- Support this for presigned URLs.  Looks like a matter of moving the new code up a few lines, see inline comment.
- Add a test for this support :-)

vishalsanfran commented:
This is ready for review @arielshaqed

N-o-Z commented on pkg/api/controller.go:
```diff
@@ -4197,6 +4197,14 @@ func (c *Controller) GetObject(w http.ResponseWriter, r *http.Request, repositor
+	etag := httputil.ETag(entry.Checksum)
+
+	// check ETag if not modified in request
+	if *params.IfNoneMatch == etag {
```
Use the swag.StringValue, this pointer can be nil

N-o-Z commented on pkg/api/controller.go:
```diff
@@ -4213,8 +4221,8 @@ func (c *Controller) GetObject(w http.ResponseWriter, r *http.Request, repositor
 		return
 	}
 
+	// setup response
```
Redundant comment

N-o-Z commented on pkg/api/controller_test.go:
```diff
@@ -2687,6 +2687,43 @@ func TestController_ObjectsGetObjectHandler(t *testing.T) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		if resp.HTTPResponse.StatusCode != http.StatusOK {
```
You can use require.NotNil(resp.HTTP200) instead of this if clause

vishalsanfran replied:
Sure, but wouldn't it be even better to do `require.Equal(t, http.StatusOK, respList.StatusCode())`

N-o-Z commented on pkg/api/controller_test.go:
```diff
@@ -2687,6 +2687,43 @@ func TestController_ObjectsGetObjectHandler(t *testing.T) {
+			t.Errorf("expected 37 bytes in content length, got back %d", resp.HTTPResponse.ContentLength)
+		}
+		etag := resp.HTTPResponse.Header.Get("ETag")
+		if etag != `"3c4838fe975c762ee97cf39fbbe566f1"` {
```
I know this has been used like this before, but let's make this expected etag a const and use it wherever this string is used literally. This will help us make the tests more robust

N-o-Z commented on pkg/api/controller_test.go:
```diff
@@ -2687,6 +2687,43 @@ func TestController_ObjectsGetObjectHandler(t *testing.T) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		if resp.HTTPResponse.StatusCode != http.StatusNotModified {
```
Same as previous comment

N-o-Z commented on api/swagger.yml:
```diff
@@ -3850,6 +3850,13 @@ paths:
+        - in: header
+          name: If-None-Match
+          description: Returns response only if the object does not have a matching ETag
+          example: "abc123efg"
```
Can we use and examples the resembles an etag a bit more

N-o-Z reviewed (changes requested):
Thank you @vishalsanfran very useful addition.
One major comment and several small ones

arielshaqed reviewed (approved):
Thanks for addressing all comments!  @N-o-Z makes good points; please address them so that we can pull.

vishalsanfran commented:
@N-o-Z addressed the PR comments, also cleaned up the unit test assertions. 
I future, I think we can apply the `require` style assertions in other tests to make the test file easier to read and less verbose.

N-o-Z reviewed (approved):
Thank you for the awesome changes!

N-o-Z commented:
> @N-o-Z addressed the PR comments, also cleaned up the unit test assertions. I future, I think we can apply the `require` style assertions in other tests to make the test file easier to read and less verbose.

Agreed. This is something that we do gradually with our tests.

vishalsanfran commented:
I'm not sure why `Docs Preview and Link Check / Docs PR - Publish preview and check links` is failing.
I cant tell the issue from the logs.
can one of you help ? @arielshaqed @N-o-Z

N-o-Z commented:
@vishalsanfran you can ignore the docs preview failure its unrelated

arielshaqed commented:
Pulling by administrative force: unrelated docs gen failure, and remaining integration tests won't run for a fork.

@vishalsanfran thanks **a lot** for this PR, and for you patience with it!
